test/compress/pure_funcs.js

```diff
@@ -293,3 +293,49 @@ unary: {
         bar();
     }
 }
+
+issue_2629_1: {
+    options = {
+        side_effects: true,
+    }
+    input: {
+        /*@__PURE__*/ f1();
+        (/*@__PURE__*/ f2)();
+        /*@__PURE__*/ (f3());
+        (/*@__PURE__*/ f4());
+        /*@__PURE__*/ f5(1)(2)(3);
+        (/*@__PURE__*/ f6(1)(2)(3));
+        /*@__PURE__*/ f7.x(1).y(2).z(3);
+        (/*@__PURE__*/ f8.x(1).y(2).z(3));
+    }
```

Comment:
Could you please add arguments with side effects for each of these calls?

We want to show that the final call's arguments with side effects are retained if the pure annotated call is dropped. Likewise, we want to demonstrate that the `AST_Call.expression` will be dropped entirely even if it happens to have side effects.

Basically we want tests to cement the defacto v3.2.2 pure annotation behavior from future changes. This would serve as the pure annotation "spec".

Reply:
I have no idea what the spec should be, since I don't use pure annotation - please list them out and I'll put those into the test cases :wink: 

This PR is an exercise in cleaning up the mess with comments - by laying a more solid foundations so at least others can make incremental improvements as there will still be cases where comments will be dropped.

Reply:
Unfortunately the current broken comment logic dictates much of the pure annotation logic. 

In my opinion this:
```js
/*@__PURE__*/(g() || h())( x(), y() );
( /*@__PURE__*/ (a() || b()) )( c(), d() ); 
```
should produce:
```js
x(), y();
(a() || b())(c(), d());
```
Notice the pure annotation in the second line marks a non-call expression so it should do nothing. v3.2.2 does not support pure annotations other than calls.

Reply:
So the first line works as described, while the second line produces `c(), d();`